armi/reactor/converters/axialExpansion/axialExpansionChanger.py

```diff
@@ -0,0 +1,396 @@
+        self.expansionData.computeThermalExpansionFactors()
+        self.axiallyExpandAssembly()
+
+    def reset(self):
+        self.linked = None
+        self.expansionData = None
+
+    def setAssembly(self, a, setFuel=True, expandFromTinputToThot=False):
+        """Set the armi assembly to be changed and init expansion data class for assembly.
+
+        Parameters
+        ----------
+         a : :py:class:`Assembly <armi.reactor.assemblies.Assembly>`
+            ARMI assembly to be changed
+        setFuel : boolean, optional
+            Boolean to determine whether or not fuel blocks should have their target components set
+            This is useful when target components within a fuel block need to be determined on-the-fly.
+        expandFromTinputToThot: bool
+            determines if thermal expansion factors should be calculated from c.inputTemperatureInC
+            to c.temperatureInC (True) or some other reference temperature and c.temperatureInC (False)
+
+        Notes
+        -----
+        When considering thermal expansion, if there is an axial temperature distribution on the assembly,
+        the axial expansion methodology will NOT perfectly preseve mass. The magnitude of the gradient of
+        the temperature distribution is the primary factor in determining the cumulative loss of mass conservation.
+        Additional details will be documented in :ref:`axialExpansion` of the documentation.
+        """
+        self.linked = AssemblyAxialLinkage(a)
+        self.expansionData = ExpansionData(
+            a, setFuel=setFuel, expandFromTinputToThot=expandFromTinputToThot
+        )
+        self._isTopDummyBlockPresent()
+
+    def applyColdHeightMassIncrease(self):
+        """
+        Increase component mass because they are declared at cold dims.
+
+        Notes
+        -----
+        A cold 1 cm tall component will have more mass that a component with the
+        same mass/length as a component with a hot height of 1 cm. This should be
+        called when the setting `inputHeightsConsideredHot` is used. This adjusts
+        the expansion factor applied during applyMaterialMassFracsToNumberDensities.
+        """
+        for c in self.linked.a.getComponents():
+            axialExpansionFactor = 1.0 + c.material.linearExpansionFactor(
+                c.temperatureInC, c.inputTemperatureInC
+            )
+            c.changeNDensByFactor(axialExpansionFactor)
+
+    def _isTopDummyBlockPresent(self):
+        """Determines if top most block of assembly is a dummy block.
+
+        Notes
+        -----
+        - If true, then axial expansion will be physical for all blocks.
+        - If false, the top most block in the assembly is artificially chopped
+          to preserve the assembly height. A runLog.Warning also issued.
+        """
+        blkLst = self.linked.a.getBlocks()
+        if not blkLst[-1].hasFlags(Flags.DUMMY):
+            runLog.warning(
```

Comment:
Do we have a unit test that covers the warnings and errors in this method?

(The primary cause of code coverage being low in ARMI right now is people adding a near infinite number of warnings, but not testing them. In the future, I am wondering if all need warnings and exceptions should be covered by tests. Though I haven't made that rule yet, obviously.)

Reply:
These should have testing, yes. I made a concerted effort a year or two ago to make sure that each of these had coverage and that axial expansion had 100% testing coverage. It should be the same...!